Fragment memtable range tombstone in the write path

cbi42 opened the pull request:
Summary:
- Right now each read fragments the memtable range tombstones #4808. This PR explores the idea of fragmenting memtable range tombstones in the write path and reads can just read this cached fragmented tombstone without any fragmenting cost. This PR only does the caching for immutable memtable, and does so right before a memtable is added to an immutable memtable list. The fragmentation is done without holding mutex to minimize its performance impact.
- db_bench is updated to print out the number of range deletions executed if there is any.

Test plan: 
- CI, added asserts in various places to check whether a fragmented range tombstone list should have been constructed.
- Benchmark: as this PR only optimizes immutable memtable path, the number of writes in the benchmark is chosen such  an immutable memtable is created and range tombstones are in that memtable.

```
single thread:
./db_bench --benchmarks=fillrandom,readrandom --writes_per_range_tombstone=1 --max_write_buffer_number=100 --min_write_buffer_number_to_merge=100 --writes=500000 --reads=100000 --max_num_range_tombstones=100

multi_thread
./db_bench --benchmarks=fillrandom,readrandom --writes_per_range_tombstone=1 --max_write_buffer_number=100 --min_write_buffer_number_to_merge=100 --writes=15000 --reads=20000 --threads=32 --max_num_range_tombstones=100
```
Commit 99cdf16464a057ca44de2f747541dedf651bae9e is included in benchmark result. It was an earlier attempt where tombstones are fragmented for each write operation. Reader threads share it using a shared_ptr which would slow down multi-thread read performance as seen in benchmark results. 
Results are averaged over 5 runs.

Single thread result:
| Max # tombstones  | main fillrandom micros/op | 99cdf16464a057ca44de2f747541dedf651bae9e | Post PR | main readrandom micros/op |  99cdf16464a057ca44de2f747541dedf651bae9e | Post PR | 
| ------------- | ------------- |------------- |------------- |------------- |------------- |------------- |
| 0    |6.68     |6.57     |6.72     |4.72     |4.79     |4.54     |
| 1    |6.67     |6.58     |6.62     |5.41     |4.74     |4.72     |
| 10   |6.59     |6.5      |6.56     |7.83     |4.69     |4.59     |
| 100  |6.62     |6.75     |6.58     |29.57    |5.04     |5.09     |
| 1000 |6.54     |6.82     |6.61     |320.33   |5.22     |5.21     |

32-thread result: note that "Max # tombstones" is per thread.
| Max # tombstones  | main fillrandom micros/op | 99cdf16464a057ca44de2f747541dedf651bae9e | Post PR | main readrandom micros/op |  99cdf16464a057ca44de2f747541dedf651bae9e | Post PR | 
| ------------- | ------------- |------------- |------------- |------------- |------------- |------------- |
| 0    |234.52   |260.25   |239.42   |5.06     |5.38     |5.09     |
| 1    |236.46   |262.0    |231.1    |19.57    |22.14    |5.45     |
| 10   |236.95   |263.84   |251.49   |151.73   |21.61    |5.73     |
| 100  |268.16   |296.8    |280.13   |2308.52  |22.27    |6.57     |

ajkr reviewed (commented):
Are we still planning to take this approach? When discussing this, I didn't realize this only takes effect when `is_range_del_table_empty_ == false`. In that case we are only adding shared_ptr access on the read path for `DeleteRange()` users, which is not ideal but also not obviously a big deal.

If you do rerun the benchmarks, can you borrow @hx235's devbig or another dual socket host (maybe `dba allocate` with some constraints)?

ajkr commented on db/memtable.cc:
```diff
@@ -661,6 +658,14 @@ Status MemTable::Add(SequenceNumber s, ValueType type,
+    auto* unfragmented_iter =
+        new MemTableIterator(*this, ReadOptions(), nullptr /* arena */,
+                             true /* use_range_del_table */);
+    std::atomic_store(&fragmented_range_tombstone_list_,
```
It seems safe to use `std::memory_order_relaxed` here since the new list won't be needed until readers start using seqnos >= this range tombstone, and we already have memory ordering on the seqno. It might not help though since I read atomic shared_ptr is often implemented with global spinlock.

ajkr replied:
This also applies to the std::atomic_load, perhaps more importantly.

ajkr commented:
For the benchmarks, perhaps 0, 1, 10, 100, and 1000 range tombstones in memtable would be good to compare. That way we'll know how it'll affect workloads that likely exist already.

cbi42 commented:
> Are we still planning to take this approach? When discussing this, I didn't realize this only takes effect when `is_range_del_table_empty_ == false`. In that case we are only adding shared_ptr access on the read path for `DeleteRange()` users, which is not ideal but also not obviously a big deal.
> 
> If you do rerun the benchmarks, can you borrow @hx235's devbig or another dual socket host (maybe `dba allocate` with some constraints)?

Yes, I still plan to work on this. I want to explore using the SuperVersion APIs, at least for the immutable memtables. Since the cost of installing SuperVersion is cheaper for immutable memtables as it's only done once per table. As for live memtables, the memtable V2 PR (#10308) shows a better worst case perf for writes, so it depends on whether we worry about that. I'll also run the benchmarks on a dual socket host and with the suggest number of tombstones. Thanks for proposing that.

ajkr reviewed (commented):
>  I want to explore using the SuperVersion APIs, at least for the immutable memtables. Since the cost of installing SuperVersion is cheaper for immutable memtables as it's only done once per table.

Got it. Is the idea to build the fragmented range tombstone list on the write path and then install it during memtable switch? It might be tricky to build it only when the memtable becomes immutable because that happens during an arbitrary write (whichever one is last to fill up the memtable, typically), which is not necessarily a `DeleteRange()`: https://github.com/facebook/rocksdb/blob/87649d3288094c528611fdbb25140a83482b8643/db/db_impl/db_impl_write.cc#L2187-L2191

ajkr commented on db/memtable.h:
```diff
@@ -603,7 +603,12 @@ class MemTable {
+
+  // The current fragmented range tombstones of this memtable.
+  // Updated when a new DeleteRange() is inserted into the memtable.
+  std::shared_ptr<FragmentedRangeTombstoneList>
+      fragmented_range_tombstone_list_;
```
I wonder if we should try ThreadLocalPtr or CoreLocalArray to hold references to the FragmentedRangeTombstoneList. Core-local is usually preferred nowadays. The FragmentedRangeTombstoneList references could be locally refcounted so we only mutate the shared refcount when the list changes.

I found this folly class interesting: https://github.com/facebook/folly/blob/4b930941724f3d2ed44dd9d7e389437dbf26d787/folly/concurrency/CoreCachedSharedPtr.h#L162. It uses shared_ptr for local refcounting (`holder`) of another shared_ptr. We might be able to use CoreLocalArray to do something similar. (I suspect our interface maybe should not be specific to shared_ptr; more details to come on that...)

ajkr commented on db/memtable.h:
```diff
@@ -601,9 +607,21 @@ class MemTable {
+  std::unique_ptr<FragmentedRangeTombstoneList>
+      fragmented_range_tombstone_list_;
+
+  // Denotes whether a fragmented range tombstone list is cached with the
+  // memtable. This should only be set be immutable memtables.
+  std::atomic<bool> has_range_tombstone_list_;
```
Do you see any possibility for an API (like `PrepareImmutable()` or something) that is called by `SwitchMemtable()` with lock released? I've also been thinking about madvise() with MADV_DONTNEED the unused memory at the same point.

ajkr replied:
Also regarding this particular variable, I wonder if `MemTableListVersion::AddRangeTombstoneIterators()` should pass the info about immutability down to the `MemTable::NewRangeTombstoneIterator()`.

cbi42 replied:
> PrepareImmutable()
I think an API like this is great once we do multiple things in the prepare phase (like when we add `madvise()` calls). One question about `unused memory`, what are the don't-needed memory in the immutable memtable? Will they be accessed by threads holding old SuperVersion that still treat this memtable as a live memtable?

cbi42 replied:
> Also regarding this particular variable, I wonder if `MemTableListVersion::AddRangeTombstoneIterators()` should 
pass the info about immutability down to the `MemTable::NewRangeTombstoneIterator()`.

Ah yes, good point. I think a similar change in Get/MultiGet calls are also needed and doable. Then we can get rid of this atomic flag.

ajkr reviewed (approved):
LGTM!

ajkr commented:
> > I want to explore using the SuperVersion APIs, at least for the immutable memtables. Since the cost of installing SuperVersion is cheaper for immutable memtables as it's only done once per table.
> 
> Got it. Is the idea to build the fragmented range tombstone list on the write path and then install it during memtable switch? It might be tricky to build it only when the memtable becomes immutable because that happens during an arbitrary write (whichever one is last to fill up the memtable, typically), which is not necessarily a `DeleteRange()`:

I changed my mind on this due to realizing the fragmented range tombstone list is already being built every read while the memtable was mutable, so it doesn't seem particularly concerning to do it once in the write path. Though I do still hope we figure out how to build the list in the mutable memtable during the `DeleteRange()` write in a next step :)

cbi42 commented:
> LGTM!

Thanks for the review. I've updated the PR with the atomic boolean variable `has_range_tombstone_list_` removed. The PR description is also updated with the latest benchmark.

cbi42 commented:
> Though I do still hope we figure out how to build the list in the mutable memtable during the `DeleteRange()` write in a next step :)

#10547 does the fragmentation for mutable memtables, but in the read path (the first reader does the fragmentation). Whether to do it in read or write path can be adjusted relatively easily. At least the idea of using CoreLocalArray to speed up reads is working.
